design/thin-edge-core.md

```diff
@@ -0,0 +1,278 @@
+  with sealed, single-purpose software images.
+  
+What makes thin-edge unique is its approach to reconcile these two poles
+with two levels of building blocks that combine along an IIoT specific API.
+* To maximize flexibility and interoperability,
+  thin-edge provides the tools to build an agent from independent executables
+  that interact over MQTT and HTTP using JSON messages.
+* To minimize resources and vulnerabilities,
+  thin-edge features fine-grain Rust components
+  as well as the rules to combine them into hardened executables.
+* To ease incremental developments,
+  thin-edge give the developers the freedom to combine both MQTT and Rust-based components
+  into purpose-specific agents.
+* To make feasible the interoperability of components provided by different tiers,
+  thin-edge comes with an extensible data model for IIoT -
+  cloud connectivity, telemetry data, device management, child devices, ...
+
+## Design Principles
+
+* Thin-edge provides the tools to develop an IIoT agent by assembling components
+  that have been implemented independently,
+  possibly by different vendors and using different programming languages.
+* There are two levels of components:
+  * MQTT-based components that are processes running on the local network
+    and interacting over MQTT and HTTP using JSON messages.
+  * Rust-based components that are actors running inside a process
+    and interacting over in-memory channels using statically typed messages.
+* These two levels of components serve different purposes and users:
+  * The Rust-based components and their assemblage are designed for robustness.
+    They follow strict combination rules enforced at compile time,
+    notably with compatibility checks between message consumers and producers.
+    These components provide the building blocks:
+    * connections to IoT cloud end-points,
+    * connections to various protocols (MQTT, HTTP, Modbus, OPC UA, ...),
+    * interactions with the operating systems (files, commands, software updates, ...),
+    * making sense of telemetry data (measurements, events, alarms, set points).
+  * The MQTT-based components bring the flexibility to interact
+    with external systems running on a different device or inside a container
+    and without enforcing a programming language or an operating system.
+    These components also provide the flexibility to experiment
+    and to add missing features without having to implement a full-fledged Rust component. 
+* To enable interactions between the MQTT-components, thin-edge provides an MQTT bus made of:
+    * a local MQTT server,
+    * an MQTT bridge that relays messages between the agent and the cloud end-points,
+    * an MQTT API that defines topics and message payloads
+      * to exchange telemetry data,
+      * to monitor components health,
+      * to trigger operations and monitor progress,
+    * a local HTTP server to exchange files.
```

Comment:
Limiting the local HTTP server just for file transfer is too short-term-ish for a vision document, isn't it? It can even be used to manage the local device inventory in the future, may be.

Reply:
Good point. I will make the purpose of this HTTP service broader.